keps/sig-etcd/4578-feature-gate/README.md

```diff
@@ -0,0 +1,383 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### New APIs
+
+New `--feature-gates` flag would be added to start the etcd server, with format like `--feature-gates=featureA=true,featureB=false`.
+
+New grpc and http endpoints would be added to query if a feature is enabled for the server or cluster.
+
+For grpc, new rpc type would be added to the `Maintenance` service.
+
+```proto
+service Maintenance {
+  ...
+  rpc FeatureGateStatus(FeatureGateStatusRequest) returns (FeatureGateStatusResponse) {
+    option (google.api.http) = {
+      post: "/v3/maintenance/featuregate"
+      body: "*"
+    };
+  }
+  ...
+}
+message Feature {
+  string name = 1;
+  bool enabled = 2;
+}
+
+message FeatureGateStatusRequest {
+  // if true, query if the features are enabled for the single server.
+  // otherwise, query if the features are enabled for the cluster.
+  bool isServerFeature = 1;
+  repeated string features = 2; // return all the enabled features if empty.
+}
+
+message FeatureGateStatusResponse {
+  ResponseHeader header = 1;
+  bool isServerFeature = 2;
+  repeated Feature features = 3;
+}
+```
+
+The http endpoints could look like:
+* `{server_endpoint}/featuregate?feature={featureName}` returns true if the feature is enabled for the cluster
+* `{server_endpoint}/featuregate?feature={featureName}&isServerFeature=true` returns true if the feature is enabled for the server
+
+### Feature Gate
+
+We will use the new `k8s.io/component-base/featuregate.VersionedSpecs`(introduced in [kep-4330](https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/4330-compatibility-versions)) to register and track the features along with their lifecycle at different release versions. 
+
+```go
+map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
+		},
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+The feature gates for a server can only be set with the `--feature-gates` flag during startup. We do not support dynamically changing the feature gates when the server is running.
```

Comment:
Should we support setting feature flags via configuration file as well?  Imo yes as many of our users prefer to configure etcd via configuration file.